MetaMask/metamask-mobile
# E2E Detox Android

### Development & PR Process
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

### Description

[Initial investigation](https://gist.github.com/leotm/16262cd2fba3d0097cb456134253f1e8)

**Context**
- https://github.com/MetaMask/mobile-planning/issues/658 cc/pairing with @NicolasMassart 

**Detox (locally) android.emu.debug**
Builds apk's ✅ boots emu ✅ launches app ✅ Detox idles 🤔 _even when logbox warnings closed_

<img width="1725" alt="Screenshot 2023-03-08 at 18 08 53" src="https://user-images.githubusercontent.com/1881059/223795361-a3b787a3-0026-4bd2-aa85-3aed2b331815.png">

**Detox (locally) android.emu.release.qa**
- Build ❌ _Sentry 401 error towards end of build_ ([screenshot](https://user-images.githubusercontent.com/1881059/223686057-6cd816da-ed0c-45cf-a6db-cddd24dec755.png))
- Add QAE2E $MODE to build shellscript ✅ https://github.com/MetaMask/metamask-mobile/pull/5913/commits/8b2aa002cc4dd3bad43c908822ee3ebdd4bf5cfe
- Build ❌ QA/Release keystore's missing locally
- Fixed forcing debug.keystore locally ✅
- Build ❌ react-native-payments minSDK 16 <= vanilla RN 0.66 ver [screenshot](https://user-images.githubusercontent.com/1881059/223713391-a047be78-8ca2-4647-80c6-8edc35cf1f00.png)
  - i.e. 21 (despite us bumping it to 23 in our build.gradle via a RN ReactAndroid/build.gradle [patch](https://github.com/MetaMask/metamask-mobile/blob/main/patches/react-native%2B0.66.5.patch))
- Fix build errors by patching lib minSDKs ([yikes](https://user-images.githubusercontent.com/1881059/223763878-ccdbf98d-7fa8-4311-98ba-40f2caa53855.png)) https://github.com/MetaMask/metamask-mobile/pull/5913/commits/f24f59212c1a2f4b93499e59d7377acdc13edcf7 ❌
  - Not feasible patching gazillion libs including new errors ([ReactAndroid:compileDebugAndroidTestJavaWithJavac](https://user-images.githubusercontent.com/1881059/223764413-68c8c191-6900-4592-8c3d-a6fa3b1b2285.png))
    - patch react-native-webview _as well_?
  - https://github.com/wix/Detox/issues/2712
  - others in a similar [boat](https://stackoverflow.com/questions/67021833/struggling-to-get-android-build-working-due-to-minsdkversion)
- Instead fix build errors in shellscript with `app:` module prefix to `assembleAndroidTest` gradle task ✅
https://github.com/MetaMask/metamask-mobile/pull/5913/commits/402f97013fc951eca6f3708ea33ba85b69c201e3
- DetoxRuntimeError ❌ [cannot find `emulator`](https://user-images.githubusercontent.com/1881059/223768377-9c8bd7e7-46db-46be-8f99-3bf5b0cda3c6.png)
- For now provide local emu name ✅ https://github.com/MetaMask/metamask-mobile/pull/5913/commits/13b8aa67716d05e230475bf5d4a9d9e05bf24443
- DetoxRuntimeError ❌ nasty [crash](https://user-images.githubusercontent.com/1881059/223770081-254dc7ba-afb0-40c6-b075-438f475a9182.png) `java.lang. NoSuchMethodError: No static method checkParameterIsNotNull(Ljava/lang/Object;Ljava/lang/String;)V`
  - [log](https://pastebin.com/zChyUi3S): emu doesn't support vulkan graphics :feelsgood: 
    - more info on ndk graphics [here](https://developer.android.com/ndk/guides/graphics/getting-started)
  - create Pixel_6_Pro_API_30 containing Vulkan emulated graphics ✅
- DetoxRuntimeError ❌ another nasty [crash](https://github.com/wix/Detox/blob/d9014dacb1fbf8451cdc0b9d8e9ae19897f0a019/detox/android/detox/src/full/java/com/wix/detox/Delegator.java#L34), time to upgrade Detox
- Upgrade Detox from 19.11.0 to 19.13.0 ✅ https://github.com/MetaMask/metamask-mobile/pull/5913/commits/1e3ba457a0000d2ad01efb113d3973a41cf9fc75
  - Socket Security [alert](https://github.com/MetaMask/metamask-mobile/pull/5913/checks?check_run_id=11857187944): `@SocketSecurity ignore detox@19.13.0`? since same postinstall in v[9.11.0](https://socket.dev/npm/package/detox/files/19.11.0/package.json#T1028-1057)
- DetoxRuntimeError ❌ another [crash](https://user-images.githubusercontent.com/1881059/223772402-0dedfd57-de32-4bb1-8141-e75ad5bf1796.png) `java.lang. NoClassDefFoundError: Failed resolution of: Lkotlin/jvm/internal/FunctionReferenceImp1;`
- Fix with Kotlin proguard rules ✅ https://github.com/MetaMask/metamask-mobile/pull/5913/commits/c3fa2ffec3690caf7acb3375e58588996187c2a7
- Now no crash 🎉 but Detox idles 🤔
- Test Detox v20 (https://github.com/MetaMask/metamask-mobile/pull/5913#issuecomment-1477758387)
- Check https://wix.github.io/Detox/docs/troubleshooting/running-tests/#tests-execution-hangs
  - Not getting `Detox can’t seem to connect to the test app(s)!` in debug/release, since already calling (via `init.js`) `await device.launchApp()` and can repro the msg if intentionally leaving it out - so can safely discard this being the issue
- Revert aforementioned patches (https://github.com/MetaMask/metamask-mobile/pull/5913/commits/85cd3fc3f680c48877e9f25dcad6a93e55addc00)

<img width="1724" alt="Screenshot 2023-03-08 at 18 09 48" src="https://user-images.githubusercontent.com/1881059/223795324-9fba1199-87d1-480a-9fe5-58d7d78c7cb8.png">

**Detox (BitRise) android.emu.release.qa**
Build ❌ _Keystore issue_ https://app.bitrise.io/build/74074844-c852-4c09-9c24-8192081fa1ee

_nb: Detox iOS debug/release build/run fine locally (debug: close warning box, then Detox continues fine)_

### Screenshots/Recordings

Above ^ and below

**Various Detox problems overcome**

<img width="1216" alt="Screenshot 2023-03-06 at 15 21 50" src="https://user-images.githubusercontent.com/1881059/223494487-e59aa6ca-7059-44f0-bdec-837ba00cabc0.png">

<img width="1170" alt="Screenshot 2023-03-07 at 11 58 46" src="https://user-images.githubusercontent.com/1881059/223494595-794ab578-734f-43c0-afc9-9cd4af3bc4b8.png">

<img width="1424" alt="Screenshot 2023-03-07 at 17 02 47" src="https://user-images.githubusercontent.com/1881059/223494933-48dcc8a1-e578-4634-ab28-120d4816d2b5.png">

### Issue

Progresses https://github.com/MetaMask/mobile-planning/issues/658

### Checklist

* [ ] <s>Tests are included if applicable</s>
